src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -244,24 +244,35 @@ public boolean isNative() {
     @Override
     public final Optional<MemorySegment> asOverlappingSlice(MemorySegment other) {
         final AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)Objects.requireNonNull(other);
-        if (overlaps(that)) {
-            final long offsetToThat = that.address() - this.address();
-            final long newOffset = offsetToThat >= 0 ? offsetToThat : 0;
-            return Optional.of(asSlice(newOffset, Math.min(this.byteSize() - newOffset, that.byteSize() + offsetToThat)));
+        if (overlaps(that) == 0) {
+            return Optional.empty();
         }
-        return Optional.empty();
+        final long offsetToThat = that.address() - this.address();
+        final long newOffset = offsetToThat >= 0 ? offsetToThat : 0;
+        return Optional.of(asSlice(newOffset, Math.min(this.byteSize() - newOffset, that.byteSize() + offsetToThat)));
     }
 
+    // Returns 1 if the regions overlap, otherwise 0.
     @ForceInline
-    boolean overlaps(AbstractMemorySegmentImpl that) {
+    int overlaps(AbstractMemorySegmentImpl that) {
```

Comment:
This modification provides a 10% total (i.e., for copy) performance improvement across all platforms for some likely common copy sizes like 16 and 24.

Branching is sometimes an expensive operation that can disrupt the execution pipeline. With this, we avoid some branching. On the flip side, now both the expressions are always executed (in their branchless form). All things considered, this seems to be a win performance-wise at the expense of more complex code.

So, there is a tradeoff to consider.